### DISCUSSION TITLE
Minuit Optimization failed. Estimated distance to minimum too large.

### DISCUSSION THREAD

**Discussion by ssaha1234** [scikit-hep/pyhf#1781] [Q&A] [answered]
Dear pyhf community, 
   I am running a simple mle fit as shown below, but minuit fails with the error ```Estimated distance to minimum too large.``` 
I have set the ```strategy``` to 0 and loosened the ```tolerance``` to 10000.  
I see this error when I changed some values under ```observations``` in my json file. The same code runs on a previous version my the json file, but not this edited version. 

## Code

```python
pyhf.set_backend("numpy",pyhf.optimize.minuit_optimizer(verbose = True, tolerance = 10000, strategy = 0))
def fitresults(constraints=None):

    constraints = constraints or []
    init_pars = model.config.suggested_init()
    fixed_params = model.config.suggested_fixed()
    for idx, fixed_val in constraints:
        init_pars[idx] = fixed_val
        fixed_params[idx] = True

    result, result_obj = pyhf.infer.mle.fit(
        data,
        model,
    	maxiter=1000000,    
        init_pars=init_pars,
        fixed_params=fixed_params,
        return_uncertainties=True,
        return_result_obj=True
    )
    bestfit = result[:,0]
    errors = result[:,1]
    return bestfit, errors, result_obj.corr

```

## Error
   
```pytb
result, result_obj = pyhf.infer.mle.fit(
  File "/afs/cern.ch/user/s/ssaha/.local/lib/python3.8/site-packages/pyhf/infer/mle.py", line 131, in fit
    return opt.minimize(
  File "/afs/cern.ch/user/s/ssaha/.local/lib/python3.8/site-packages/pyhf/optimize/mixins.py", line 184, in minimize
    result = self._internal_minimize(
  File "/afs/cern.ch/user/s/ssaha/.local/lib/python3.8/site-packages/pyhf/optimize/mixins.py", line 64, in _internal_minimize
    raise exceptions.FailedMinimization(result)
pyhf.exceptions.FailedMinimization: Optimization failed. Estimated distance to minimum too large.
```

Is there a way to debug this in a better way ? I can provide the full script if needed. 
Thanks a lot !

Best, 
Shreya

**Comment by matthewfeickert** [comment]
@ssaha1234 A question about your environment so that we know what versions we're dealing with here. I see from your traceback that you're on LXPLUS at CERN but that you're also not in a Python virtual environment. I would highly encourage you to do everything you can to only work inside of a Python virtual environment where you install the packages yourself, else you can run into lots of problems. For example, I assume here that you have an LCG view setup as well? If so, then it is very hard to know what versions of libraries you actually have to work with.

Can you tell us what your environment is like? A good start would be the output of

```console
$ python -m pip list | grep 'pyhf\|scipy\|numpy\|iminuit'
```

---

Aside: If you're trying to work with LCG views on a CVMFS system it can be very challenging to work with virtual environments, but it is possible! I've started this toy implementation of a little project that abstracts away some of the difficulty here so you can use a Python virtual environment almost like you normally would regardless of the fact that you're inside of an LCG view: https://github.com/matthewfeickert/cvmfs-venv. If you try it out and have questions / run into problems please ping me on its GitHub Issues.

**Reply by ssaha1234** [reply]
Hi @matthewfeickert, 
     This is the output I get for ```$ python -m pip list | grep 'pyhf\|scipy\|numpy\|iminuit'``` - 

```
iminuit                            2.7.0
numpy                              1.22.2
pyhf                               0.6.3
root-numpy                         4.8.0
scipy                              1.5.1
```
Thank you very much for the link to the project. I will try it out in the near future :) 

I have a slightly unrelated Q about using ```jax```.  For example, the fit I am trying out takes ~ 1.5 hours, is there a way to use ```jax``` to speed it up ? I tried it yesterday by setting ```python.set_backend("jax")```, but I ran into quite a few errors (I can provide details, if that helps). Is there a way to implement it in the pyhf version I am using ?

Thanks ! 
Best, 
Shreya

**Comment by alexander-held** [comment, marked as answer]
Hi @ssaha1234, the most common cause for this in my experience is that your fit model lacks relevant degrees of freedom to model the data you are fitting.

> I see this error when I changed some values under `observations` in my json file.

This also points towards the explanation: by sufficiently changing numbers there you can essentially make any fit fail, because the observation will end up as something that cannot be described by the model anymore.

I would suggest plotting your distributions to see how close your observations are to the model. When also visualizing model uncertainties, you can roughly judge whether some observation may be so far away that a fit might fail. To visualize, you can use `model.expected_data` to get the relevant distributions. Alternatively, [`cabinetry`](https://cabinetry.readthedocs.io/) provides this functionality: [`cabinetry.model_utils.prediction`](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.model_utils.prediction) will give you the relevant information, which you can then use with [`cabinetry.visualize.data_mc`](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.visualize.data_mc) to obtain plots. See e.g. the notebooks in [cabinetry/cabinetry-tutorials](https://github.com/cabinetry/cabinetry-tutorials) for examples.

**Reply by ssaha1234** [reply]
Hi @alexander-held, 
   Yes, I would have guessed the same, my new observation values are causing this fit error. Thanks for the link to the tutorial, I am going to make some intermediate plots to see, what is going wrong. I re ran the fit with my old observation values and it works just fine with the nominal tolerance and strategy. 
Thanks ! 
Best, 
Shreya

**Reply by ssaha1234** [reply]
Just an update on this, I had to change a few more values in the likelihood file, and now the fit works ! 
Thanks !
Best, 
Shreya
